runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunnerInfo.java

```diff
@@ -19,23 +19,66 @@
 
 import static com.google.common.base.Preconditions.checkState;
 
+import com.google.common.collect.ImmutableMap;
 import java.io.IOException;
 import java.io.InputStream;
+import java.util.Map;
 import java.util.Properties;
+import org.apache.beam.sdk.util.ReleaseInfo;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * Populates versioning and other information for {@link DataflowRunner}.
  */
-public final class DataflowRunnerInfo {
+public final class DataflowRunnerInfo extends ReleaseInfo {
   private static final Logger LOG = LoggerFactory.getLogger(DataflowRunnerInfo.class);
 
-  private static final String PROPERTIES_PATH =
+  private static final String APACHE_BEAM_DISTRIBUTION_PROPERTIES_PATH =
       "/org/apache/beam/runners/dataflow/dataflow.properties";
+  private static final String DATAFLOW_DISTRIBUTION_PROPERTIES_PATH =
+      "/org/apache/beam/runners/dataflow/dataflow-distribution.properties";
+  private static final String FNAPI_ENVIRONMENT_MAJOR_VERSION_KEY =
+      "fnapi.environment.major.version";
+  private static final String LEGACY_ENVIRONMENT_MAJOR_VERSION_KEY =
+      "legacy.environment.major.version";
+  private static final String CONTAINER_VERSION_KEY = "container.version";
 
   private static class LazyInit {
-    private static final DataflowRunnerInfo INSTANCE = new DataflowRunnerInfo(PROPERTIES_PATH);
+    private static final DataflowRunnerInfo INSTANCE;
+    static {
+      Properties properties;
+      try {
+        properties = load(DATAFLOW_DISTRIBUTION_PROPERTIES_PATH);
+        if (properties == null) {
+          properties = load(APACHE_BEAM_DISTRIBUTION_PROPERTIES_PATH);
+        }
+        if (properties == null) {
+          // Print a warning if we can not load either the Dataflow distribution properties
+          // or the
+          LOG.warn("Dataflow runner properties resource not found.");
+          properties = new Properties();
+        }
+      } catch (IOException e) {
+        LOG.warn("Error loading Dataflow runner properties resource: ", e);
+        properties = new Properties();
+      }
+
+      // Inherit the name and version from the Apache Beam distribution if this isn't
+      // the Dataflow distribution.
+      if (!properties.containsKey("name")) {
+        properties.setProperty("name", ReleaseInfo.getReleaseInfo().getName());
```

Comment:
<!--new_thread; commit:e9176e9939f0dd61789718203cca95ed760e0954; resolved:0-->
Seems like this should be additive, not necessarily an overwrite.

Reply:
Dataflow service currently relies on this property to detect deprecated SDKs